compiler/noirc_frontend/src/elaborator/types.rs

```diff
@@ -412,20 +410,57 @@ impl<'context> Elaborator<'context> {
                 let kind = self
                     .interner
                     .get_global_let_statement(id)
-                    .map(|let_statement| Kind::numeric(let_statement.r#type))
+                    .map(|let_statement| {
+                        if let_statement.r#type == Type::FieldElement {
+                            Kind::IntegerOrField
+                        } else {
+                            Kind::numeric(let_statement.r#type)
+                        }
+                    })
                     .unwrap_or(Kind::u32());
 
-                // TODO(https://github.com/noir-lang/noir/issues/6238):
-                // support non-u32 generics here
-                if !kind.unifies(&Kind::u32()) {
-                    let error = TypeCheckError::EvaluatedGlobalIsntU32 {
-                        expected_kind: Kind::u32().to_string(),
-                        expr_kind: kind.to_string(),
-                        expr_span: path.span(),
-                    };
-                    self.push_err(error);
-                }
-                Some(Type::Constant(self.eval_global_as_array_length(id, path).into(), kind))
+                let Some(stmt) = self.interner.get_global_let_statement(id) else {
+                    if let Some(global) = self.unresolved_globals.remove(&id) {
+                        self.elaborate_global(global);
+                        return self.lookup_generic_or_global_type(path);
+                    } else {
+                        let path = path.clone();
+                        self.push_err(ResolverError::NoSuchNumericTypeVariable { path });
+                        return None;
+                    }
+                };
+
+                let rhs = stmt.expression;
+                let span = self.interner.expr_span(&rhs);
+
+                let Some(global_value) = &self.interner.get_global(id).value else {
+                    self.push_err(ResolverError::UnevaluatedGlobalType { span });
+                    return None;
+                };
+
+                let Some(global_value) = global_value.to_field_element() else {
+                    let global_value = global_value.clone();
+                    if global_value.is_integral() {
+                        self.push_err(ResolverError::NegativeGlobalType { span, global_value });
+                    } else {
+                        self.push_err(ResolverError::NonIntegralGlobalType { span, global_value });
+                    }
+                    return None;
+                };
+
+                let global_value = match kind.ensure_value_fits(global_value, span) {
+                    Ok(global_value) => global_value,
+                    Err(_err) => {
+                        self.push_err(ResolverError::GlobalLargerThanKind {
```

Comment:
Keeping with the pattern above, this could be `let Ok(global_value) = kind.ensure_value_fits(...) else {` since the error is unused.